Apple: Refactor deployment target version parsing

madsmtm opened the pull request:
Refactor deployment target parsing to make it easier to do https://github.com/rust-lang/rust/pull/129342 (I wanted to make sure of all the places that `std::env::var` is called).

Specifically, my goal was to minimize the amount of target-specific configuration, so to that end I renamed the `opts` function that generates the `TargetOptions` to `base`, and made it return the LLVM target and `target_arch` too. In the future, I would like to move even more out of the target files and into `spec::apple`, as it makes it easier for me to maintain.

For example, this fixed a bug in `aarch64-apple-watchos`, which wasn't passing the deployment target as part of the LLVM triple. This (probably) fixes https://github.com/rust-lang/rust/issues/123582 and fixes https://github.com/rust-lang/rust/issues/107630.

We also now parse the patch version of deployment targets, allowing the user to specify e.g. `MACOSX_DEPLOYMENT_TARGET=10.12.6`.

Finally, this fixes the LLVM target name for visionOS, it should be `*-apple-xros` and not `*-apple-visionos`.

Since I have changed all the Apple targets here, I smoke-tested my changes by running the following:
```console
# Build each target
./x build library --target="aarch64-apple-darwin,aarch64-apple-ios,aarch64-apple-ios-macabi,aarch64-apple-ios-sim,aarch64-apple-tvos,aarch64-apple-tvos-sim,aarch64-apple-visionos,aarch64-apple-visionos-sim,aarch64-apple-watchos,aarch64-apple-watchos-sim,arm64_32-apple-watchos,arm64e-apple-ios,armv7k-apple-watchos,armv7s-apple-ios,i386-apple-ios,x86_64-apple-darwin,x86_64-apple-ios,x86_64-apple-ios-macabi,x86_64-apple-tvos,x86_64-apple-watchos-sim,x86_64h-apple-darwin"

# Test that we can still at least link basic projects
cargo new foobar && cd foobar && cargo +stage1 build --target=aarch64-apple-darwin --target=aarch64-apple-ios --target=aarch64-apple-ios-macabi --target=aarch64-apple-ios-sim --target=aarch64-apple-tvos --target=aarch64-apple-tvos-sim --target=aarch64-apple-visionos --target=aarch64-apple-visionos-sim --target=aarch64-apple-watchos --target=aarch64-apple-watchos-sim --target=arm64_32-apple-watchos --target=armv7s-apple-ios --target=i386-apple-ios --target=x86_64-apple-darwin --target=x86_64-apple-ios --target=x86_64-apple-ios-macabi --target=x86_64-apple-tvos --target=x86_64-apple-watchos-sim --target=x86_64h-apple-darwin
```

I couldn't build for the `arm64e-apple-darwin` target, the `armv7k-apple-watchos` and `arm64e-apple-ios` targets failed to link, and I know that the `i686-apple-darwin` target requires a bit of setup, but all of this is as it was before this PR.

r? thomcc

CC @BlackHoleFox

I would recommend using `rollup=never` when merging this, in case we need to bisect this later.

madsmtm committed:
Apple: Add comments for `-platform_version` linker argument

madsmtm commented:
@bors rollup=never

There are quite a few changes here, and we might need to bisect this later.

bors commented:
@madsmtm: :key: Insufficient privileges: not in try users

thomcc commented:
@bors rollup=never

thomcc commented on compiler/rustc_target/src/spec/base/apple/mod.rs:
```diff
@@ -230,58 +274,115 @@ pub fn platform(target: &Target) -> Option<u32> {
     };
-
-    Some((major, minor))
+    let abi = match &*target.abi {
```
It bugs me that these are ABIs (and not envs, like they are in clang) but changing that isn't practical due to compatibility.

thomcc commented on compiler/rustc_target/src/spec/base/apple/mod.rs:
```diff
@@ -230,58 +274,115 @@ pub fn platform(target: &Target) -> Option<u32> {
+        // Use 11.0 on Aarch64 as that's the earliest version with M1 support.
+        ("macos", Arch::Arm64 | Arch::Arm64e, _) => (11, 0, 0),
+        ("ios", Arch::Arm64e, _) => (14, 0, 0),
+        // Mac Catalyst defaults to 13.1 in Clang.
```
This is a behavioral change, no? Not that this is a problem, just want to make sure.

madsmtm replied:
No, we did this before too, see the previous `ios_deployment_target`

thomcc reviewed (approved):
This looks fine to me. Was a hassle to verify all the targets match what they used to.

thomcc commented:
I'm not a compiler-team or compiler-contributor so I'll roll to someone who is, but I'd have R+ed this if I were.
r? compiler

bors commented:
:umbrella: The latest upstream changes (presumably #129777) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

madsmtm commented:
r? compiler

petrochenkov commented on compiler/rustc_target/src/spec/base/apple/mod.rs:
```diff
@@ -97,53 +98,95 @@ impl TargetAbi {
+        // Same story when linking with cc, since that ends up invoking ld64.
+        I386 | I686 => "i386",
+        X86_64 => "x86_64",
+        X86_64h => "x86_64h",
```
Maybe factor out this into a function and put it next to `fn target_name` to highlight their similarities and differences?

madsmtm replied:
Yeah, kinda wanted to get rid of `target_name` too, and instead just inline it at the usage sites, I felt that was clearer. But anyhow, I've done as you've suggested.

petrochenkov commented on compiler/rustc_target/src/spec/base/apple/mod.rs:
```diff
@@ -97,53 +98,95 @@ impl TargetAbi {
 }
 
 fn pre_link_args(os: &'static str, arch: Arch, abi: TargetAbi) -> LinkArgs {
+    let mut args = LinkArgs::new();
```
I think `TargetOptions::link_args` can still be used here instead of `LinkArgs::new`.

madsmtm commented:
Thanks for the review!

@rustbot ready

BlackHoleFox reviewed (approved):
Since this has been looked at someone on the compiler team, here's my lookover from the Apple perspective since I was CCed.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10708614633/job/29691343143) [(plain)](https://github.com/rust-lang/rust/commit/d69cfd75a02c6db6c25ee88631cf055808283003/checks/29691343143/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.890 Building wheels for collected packages: reuse
#13 2.891   Building wheel for reuse (pyproject.toml): started
#13 3.146   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.148   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.148   Stored in directory: /tmp/pip-ephem-wheel-cache-0dss_vfi/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.151 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.560 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.560 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.138 Collecting virtualenv
#13 4.138 Collecting virtualenv
#13 4.177   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.249      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 81.3 MB/s eta 0:00:00
#13 4.294 Collecting distlib<1,>=0.3.7
#13 4.298   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.305      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 93.3 MB/s eta 0:00:00
#13 4.345 Collecting filelock<4,>=3.12.2
#13 4.349   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.384 Collecting platformdirs<5,>=3.9.1
#13 4.389   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.471 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.663 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.7s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      210880 kB
DirectMap2M:     8177664 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/842d6fc32e3d0d26bb11fbe6a2f6ae2afccc06cb/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-842d6fc32e3d0d26bb11fbe6a2f6ae2afccc06cb-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
     //
     // The linker has heuristics to determine the desired architecture, but to
     // be safe, and to avoid a warning, we set the architecture explicitly.
-    let mut args = TargetOptions::link_args(
-        LinkerFlavor::Darwin(Cc::No, Lld::No),
-        &["-arch", arch.ld_arch()],
+    let mut args =
+    let mut args =
+        TargetOptions::link_args(LinkerFlavor::Darwin(Cc::No, Lld::No), &["-arch", arch.ld_arch()]);
 
     // From the man page for ld64 (`man ld`):
     // > This is set to indicate the platform, oldest supported version of
Diff in /checkout/compiler/rustc_target/src/spec/base/apple/mod.rs:185:
     if abi != TargetAbi::MacCatalyst {
         // CC forwards the `-arch` to the linker, so we use the same value
         // here intentionally.
-        add_link_args(&mut args, LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-arch", arch.ld_arch()]);
+            &mut args,
+            &mut args,
+            LinkerFlavor::Darwin(Cc::Yes, Lld::No),
+            &["-arch", arch.ld_arch()],
     } else {
         add_link_args_iter(
             &mut args,
             &mut args,
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/alloc/src/collections/vec_deque/into_iter.rs" "/checkout/library/alloc/src/collections/vec_deque/iter_mut.rs" "/checkout/library/alloc/src/collections/vec_deque/spec_from_iter.rs" "/checkout/library/alloc/src/collections/vec_deque/iter.rs" "/checkout/library/alloc/src/collections/vec_deque/macros.rs" "/checkout/library/alloc/src/collections/vec_deque/mod.rs" "/checkout/library/alloc/src/collections/vec_deque/spec_extend.rs" "/checkout/library/alloc/src/collections/vec_deque/tests.rs" "/checkout/library/alloc/src/collections/mod.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7a_kmc_solid_asp3_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64le_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7a_uwp_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_fuchsia.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsel_sony_psp.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_apple_ios_sim.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_trusty.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsel_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/crt_objects.rs" "/checkout/compiler/rustc_target/src/spec/mod.rs" "/checkout/library/alloc/src/collections/binary_heap/mod.rs" "/checkout/library/alloc/src/collections/binary_heap/tests.rs" "/checkout/library/alloc/src/collections/linked_list.rs" "/checkout/compiler/rustc_target/src/spec/tests/tests_impl.rs" "/checkout/library/alloc/src/collections/linked_list/tests.rs" "/checkout/library/alloc/src/collections/btree/set.rs" "/checkout/library/alloc/src/collections/btree/fix.rs" "/checkout/library/alloc/src/collections/btree/remove.rs" "/checkout/library/alloc/src/collections/btree/mem.rs" "/checkout/library/alloc/src/collections/btree/search.rs" "/checkout/library/alloc/src/collections/btree/node.rs" "/checkout/library/alloc/src/collections/btree/dedup_sorted_iter.rs" "/checkout/library/alloc/src/collections/btree/borrow.rs" "/checkout/library/alloc/src/collections/btree/merge_iter.rs" "/checkout/library/alloc/src/collections/btree/set/tests.rs" "/checkout/library/alloc/src/collections/btree/mod.rs" "/checkout/library/alloc/src/collections/btree/set_val.rs" "/checkout/library/alloc/src/collections/btree/navigate.rs" "/checkout/library/alloc/src/collections/btree/split.rs" "/checkout/library/alloc/src/collections/btree/map/entry.rs" "/checkout/library/alloc/src/collections/btree/map/tests.rs" "/checkout/library/alloc/src/collections/btree/map.rs" "/checkout/library/alloc/src/collections/btree/append.rs" "/checkout/library/alloc/src/collections/btree/node/tests.rs" "/checkout/compiler/rustc_target/src/spec/base/android.rs" "/checkout/library/alloc/src/collections/btree/borrow/tests.rs" "/checkout/library/alloc/src/macros.rs" "/checkout/compiler/rustc_target/src/spec/base/apple/mod.rs" "/checkout/compiler/rustc_target/src/spec/base/apple/tests.rs" "/checkout/compiler/rustc_target/src/spec/base/hurd.rs" "/checkout/compiler/rustc_target/src/spec/base/illumos.rs" "/checkout/compiler/rustc_target/src/spec/base/msvc.rs" "/checkout/compiler/rustc_target/src/spec/base/redox.rs" "/checkout/compiler/rustc_target/src/spec/base/hurd_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/thumb.rs" "/checkout/compiler/rustc_target/src/spec/base/linux_ohos.rs" "/checkout/compiler/rustc_target/src/spec/base/openbsd.rs" "/checkout/compiler/rustc_target/src/spec/base/aix.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_uwp_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/nto_qnx.rs" "/checkout/compiler/rustc_target/src/spec/base/teeos.rs" "/checkout/compiler/rustc_target/src/spec/base/dragonfly.rs" "/checkout/compiler/rustc_target/src/spec/base/wasm.rs" "/checkout/library/alloc/src/collections/vec_deque/drain.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Sep  4 20:03:03 UTC 2024
  network time: Wed, 04 Sep 2024 20:03:03 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10713660255/job/29706067461) [(plain)](https://github.com/rust-lang/rust/commit/b202b0b46b6dc3a315de378e826279eef63efaf8/checks/29706067461/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.820 Building wheels for collected packages: reuse
#13 2.821   Building wheel for reuse (pyproject.toml): started
#13 3.065   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.066   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.066   Stored in directory: /tmp/pip-ephem-wheel-cache-v9tgkkpg/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.069 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.461 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.461 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.977 Collecting virtualenv
#13 3.977 Collecting virtualenv
#13 4.023   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.232      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 27.5 MB/s eta 0:00:00
#13 4.274 Collecting distlib<1,>=0.3.7
#13 4.282   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.294      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 43.0 MB/s eta 0:00:00
#13 4.330 Collecting filelock<4,>=3.12.2
#13 4.337   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.370 Collecting platformdirs<5,>=3.9.1
#13 4.377   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.458 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.644 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.7s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      188352 kB
DirectMap2M:     8200192 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/009e73825af0e59ad4fc603562e038b3dbd6593a/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-009e73825af0e59ad4fc603562e038b3dbd6593a-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
     //
     // The linker has heuristics to determine the desired architecture, but to
     // be safe, and to avoid a warning, we set the architecture explicitly.
-    let mut args = TargetOptions::link_args(
-        LinkerFlavor::Darwin(Cc::No, Lld::No),
-        &["-arch", arch.ld_arch()],
+    let mut args =
+    let mut args =
+        TargetOptions::link_args(LinkerFlavor::Darwin(Cc::No, Lld::No), &["-arch", arch.ld_arch()]);
 
     // From the man page for ld64 (`man ld`):
     // > This is set to indicate the platform, oldest supported version of
Diff in /checkout/compiler/rustc_target/src/spec/base/apple/mod.rs:173:
     if abi != TargetAbi::MacCatalyst {
         // CC forwards the `-arch` to the linker, so we use the same value
         // here intentionally.
-        add_link_args(&mut args, LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-arch", arch.ld_arch()]);
+            &mut args,
+            &mut args,
+            LinkerFlavor::Darwin(Cc::Yes, Lld::No),
+            &["-arch", arch.ld_arch()],
     } else {
         add_link_args_iter(
             &mut args,
             &mut args,
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/tests/codegen/issues/issue-68667-unwrap-combinators.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv8r_none_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv6_unknown_netbsd_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm_unknown_linux_gnueabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_apple_watchos_sim.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_none_softfloat.rs" "/checkout/compiler/rustc_target/src/spec/targets/xtensa_esp32s2_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_be_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imac_esp_espidf.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7a_kmc_solid_asp3_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/i586_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm64_32_apple_watchos.rs" "/checkout/compiler/rustc_target/src/spec/targets/i586_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv6m_nuttx_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsel_unknown_linux_uclibc.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7a_kmc_solid_asp3_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64le_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7a_uwp_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_fuchsia.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsel_sony_psp.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_apple_ios_sim.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_trusty.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsel_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/crt_objects.rs" "/checkout/compiler/rustc_target/src/spec/mod.rs" "/checkout/compiler/rustc_target/src/spec/tests/tests_impl.rs" "/checkout/compiler/rustc_target/src/spec/base/android.rs" "/checkout/compiler/rustc_target/src/spec/base/apple/mod.rs" "/checkout/compiler/rustc_target/src/spec/base/apple/tests.rs" "/checkout/compiler/rustc_target/src/spec/base/hurd.rs" "/checkout/compiler/rustc_target/src/spec/base/illumos.rs" "/checkout/compiler/rustc_target/src/spec/base/msvc.rs" "/checkout/compiler/rustc_target/src/spec/base/redox.rs" "/checkout/compiler/rustc_target/src/spec/base/hurd_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/thumb.rs" "/checkout/compiler/rustc_target/src/spec/base/linux_ohos.rs" "/checkout/compiler/rustc_target/src/spec/base/openbsd.rs" "/checkout/compiler/rustc_target/src/spec/base/aix.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_uwp_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/nto_qnx.rs" "/checkout/compiler/rustc_target/src/spec/base/teeos.rs" "/checkout/compiler/rustc_target/src/spec/base/dragonfly.rs" "/checkout/compiler/rustc_target/src/spec/base/wasm.rs" "/checkout/compiler/rustc_target/src/spec/base/solaris.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_uwp_msvc.rs" "/checkout/compiler/rustc_target/src/spec/base/avr_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/linux.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/netbsd.rs" "/checkout/compiler/rustc_target/src/spec/base/haiku.rs" "/checkout/compiler/rustc_target/src/spec/base/solid.rs" "/checkout/compiler/rustc_target/src/spec/base/linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/mod.rs" "/checkout/compiler/rustc_target/src/spec/base/uefi_msvc.rs" "/checkout/compiler/rustc_target/src/spec/base/xtensa.rs" "/checkout/compiler/rustc_target/src/spec/base/bpf.rs" "/checkout/compiler/rustc_target/src/spec/base/freebsd.rs" "/checkout/compiler/rustc_target/src/spec/base/fuchsia.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_gnullvm.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_freebsd.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Thu Sep  5 04:11:06 UTC 2024
  network time: Thu, 05 Sep 2024 04:11:06 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

madsmtm committed:
Apple: Refactor deployment target version parsing

- Merge minimum OS version list into one function (makes it easier to
  see the logic in it).
- Parse patch deployment target versions.
- Consistently specify deployment target in LLVM target (previously
  omitted on `aarch64-apple-watchos`).

madsmtm commented on compiler/rustc_target/src/spec/base/apple/mod.rs:
```diff
@@ -230,58 +274,115 @@ pub fn platform(target: &Target) -> Option<u32> {
+    // we build with a version specific `llvm_target`, with the version
+    // set high enough. Luckily one LC_BUILD_VERSION is enough, for Xcode
+    // to pick it up (since std and core are still built with the fallback
+    // of version 7.0 and hence emit the old LC_IPHONE_MIN_VERSION).
```
I've rewritten the comment and moved it to `compiler/rustc_codegen_ssa/src/back/metadata.rs` where I felt the information was more relevant.

madsmtm committed:
Apple: Improve comments for `-arch` linker argument

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit bd56857e31e9324d867a4d3184301a1ef510d316 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov bd56857e31e9324d867a4d3184301a1ef510d316 -->
<!-- homu: {"type":"Approved","sha":"bd56857e31e9324d867a4d3184301a1ef510d316","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit bd56857e31e9324d867a4d3184301a1ef510d316 with merge 9afe7136958edaa403f0b0eb00f0353c125b7352...
<!-- homu: {"type":"BuildStarted","head_sha":"bd56857e31e9324d867a4d3184301a1ef510d316","merge_sha":"9afe7136958edaa403f0b0eb00f0353c125b7352"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10747556832/job/29812198515)
Approved by: petrochenkov
Pushing 9afe7136958edaa403f0b0eb00f0353c125b7352 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10747556832/job/29812198515"},"merge_sha":"9afe7136958edaa403f0b0eb00f0353c125b7352"} -->

rust-timer commented:
Finished benchmarking commit (9afe7136958edaa403f0b0eb00f0353c125b7352): [comparison URL](https://perf.rust-lang.org/compare.html?start=26b5599e4d6ed2b45152c60493c1788c0a27533d&end=9afe7136958edaa403f0b0eb00f0353c125b7352&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=26b5599e4d6ed2b45152c60493c1788c0a27533d&end=9afe7136958edaa403f0b0eb00f0353c125b7352&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=26b5599e4d6ed2b45152c60493c1788c0a27533d&end=9afe7136958edaa403f0b0eb00f0353c125b7352&stat=max-rss)
<details>
<summary>Results (primary 2.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.7% | [2.7%, 2.7%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.7% | [2.7%, 2.7%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=26b5599e4d6ed2b45152c60493c1788c0a27533d&end=9afe7136958edaa403f0b0eb00f0353c125b7352&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=26b5599e4d6ed2b45152c60493c1788c0a27533d&end=9afe7136958edaa403f0b0eb00f0353c125b7352&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 756.217s -> 756.798s (0.08%)
**Artifact size**: 341.14 MiB -> 341.03 MiB (-0.03%)
